feat: Convert c_bind apis to new msgpack apis. Potential fast IPC to bb in future by including native code.

charlielye opened the pull request:
* A lot of files touched here, but many are deletions (note additions vs removals --->).
* Add all c_bind apis, to the new msgpack api command handler.
* Add Apple support for shm ipc.
* NAPI module now supports msgpack entrypoint, via which you can call api methods.
* Erased the old c bindgen tool.
* Reworked the Barretenberg (async) and BarretenbergSync classes. They take options, one of which is an optional backend type. If not given will try and pick the best backend. Backends include:
  * Wasm
  * WasmWorker
  * NativeUnixDomainSocket
  * NativeSharedMemory
* There's a benchmark class in bb.js poseidon that measures cost for various backends. See results below. WASM is using the new msgpack api, Direct WASM is using the old cbind api (so you can see msgpack overhead).
* Only WASM and shared memory backends are supported by BarretenbergSync.
* All `foundation/crypto` primitives updated to use new api.
* **This PR does NOT enable native code.** To enable will require some ci efforts to copy `bb` and napi modules to `ts/build/<arch>-<os>` folder. Will be done in a follow up PR.

```
┌─ Size   2 field elements ──────────────────────────────────┐
│ WASM:                 491.61ms (  49.16µs/call) [baseline] │
│ Direct WASM:          438.22ms (  43.82µs/call) -  10.9%   │
│ Native Socket:        608.83ms (  60.88µs/call) +  23.8%   │
│ Native Shared:        268.16ms (  26.82µs/call) -  45.5%   │
│ Native Shared Sync:   238.86ms (  23.89µs/call) -  51.4%   │
└────────────────────────────────────────────────────────────┘
┌─ Size   4 field elements ──────────────────────────────────┐
│ WASM:                 738.71ms (  73.87µs/call) [baseline] │
│ Direct WASM:          667.25ms (  66.72µs/call) -   9.7%   │
│ Native Socket:        711.70ms (  71.17µs/call) -   3.7%   │
│ Native Shared:        418.90ms (  41.89µs/call) -  43.3%   │
│ Native Shared Sync:   383.39ms (  38.34µs/call) -  48.1%   │
└────────────────────────────────────────────────────────────┘
┌─ Size   8 field elements ──────────────────────────────────┐
│ WASM:                1066.59ms ( 106.66µs/call) [baseline] │
│ Direct WASM:          970.95ms (  97.09µs/call) -   9.0%   │
│ Native Socket:        870.64ms (  87.06µs/call) -  18.4%   │
│ Native Shared:        567.65ms (  56.77µs/call) -  46.8%   │
│ Native Shared Sync:   546.37ms (  54.64µs/call) -  48.8%   │
└────────────────────────────────────────────────────────────┘
```

dbanks12 commented on avm-transpiler/avm_transpiler.h:
The comments in this file are probably _technically_ right since any brillig program is wrapped in an ACIR program. But we don't _really_ transpile ACIR --> AVM. We transpile a contract's Brillig program for its public contract code to AVM bytecode.

ludamad replied:
Missed this, at some point AI wrote this I guess :') good catch

ludamad replied:
https://github.com/AztecProtocol/aztec-packages/pull/17829/files#diff-b159f49a4ad089507324570bdf97aa8f199f0a45a61146c23f0177cee23aa300

dbanks12 commented on barretenberg/cpp/pil/vm2/bytecode/bc_decomposition.pil:
```diff
@@ -59,11 +59,11 @@ pol commit bytes_rem_min_one_inv;
 sel * ((bytes_remaining - 1) * (last_of_contract * (1 - bytes_rem_min_one_inv) + bytes_rem_min_one_inv) + last_of_contract - 1) = 0;
 
 // The following alias is a boolean because last_of_contract == 1 ==> sel == 1 ==> first_row == 0 (sel' is defined)
-pol FIRST_OR_LAST_CONTRACT = precomputed.first_row + last_of_contract;
+pol LATCH_CONDITION = precomputed.first_row + last_of_contract;
 
```
Looks like you have some unwanted diff with latest `next`

dbanks12 commented on barretenberg/cpp/src/barretenberg/vm2/constraining/relations/bc_retrieval.test.cpp:
```diff
@@ -135,7 +135,6 @@ TEST(BytecodeRetrievalConstrainingTest, SuccessfulRetrieval)
 
     check_relation<bc_retrieval>(trace);
     check_interaction<BytecodeTraceBuilder,
-                      lookup_bc_retrieval_bytecode_hash_is_correct_settings,
                       lookup_bc_retrieval_class_id_derivation_settings,
```
Again, unwanted diffs from next

dbanks12 reviewed (commented):
The transpiler stuff looks fine to me. Looks like there is some unwanted diff from next. You probably just need to merge in next.

ludamad commented on barretenberg/cpp/src/barretenberg/bbapi/bbapi_crypto.cpp:
```diff
@@ -9,4 +18,97 @@ Poseidon2Hash::Response Poseidon2Hash::execute(BBApiRequest& request) &&
+{
+    (void)request;
+
+    // Copy ciphertext as AES decrypts in-place
```
We could just resize `ciphertext` as the style of this API is to totally consume the buffers provided. Also prefer `BB_UNUSED const BBApiRequest& request`.

ludamad commented on barretenberg/ts/scripts/build_wasm.sh:
```diff
@@ -6,7 +6,7 @@ set -e
 if [ -z "$SKIP_CPP_BUILD" ] && [ "${CI:-0}" -eq 0 ]; then
-  parallel --line-buffered --tag 'denoise "../cpp/bootstrap.sh {}"' ::: build_wasm build_wasm_threads
+  parallel --line-buffered --tag '../cpp/bootstrap.sh {}' ::: build_wasm build_wasm_threads
 fi
```
Note: I find this 'convenience call' to cpp bootstrap has only ever caused me pain. every other bootstrap pattern i'm used to just adding multiple bootstrap commands to my buffer, on the other hand I have to set an env var so this doesn't blow out things when i build with debug symbols

ludamad commented on barretenberg/ts/src/bb_backends/wasm.ts:
```diff
@@ -0,0 +1,110 @@
+ *
+ * Worker mode (default): Runs WASM on a worker thread to avoid blocking the main thread.
+ *   - Browser-safe: Won't block UI during long operations
+ *   - Overhead: ~3-4x slower due to serialize/deserialize for each call
```
super nits: just saying 3-4x slower/faster is a bit misleading. Also kinda funny to say to use the faster one for benchmarks :)

charlielye replied:
claude comments 😬 i guess the 3-4x is for something tiny like a poseidon hash.

ludamad commented:
OK I read through it all. wicked stuff! Really good for bb.js to get love like this. I liked a lot of what I saw. approving

charlielye committed:
feat: Convert c_bind apis to new msgpack apis. Potential fast IPC to bb in future by including native code.

* A lot of files touched here, but many are deletions (note additions vs removals --->).
* Add all c_bind apis, to the new msgpack api command handler.
* Add Apple support for shm ipc.
* NAPI module now supports msgpack entrypoint, via which you can call api methods.
* Erased the old c bindgen tool.
* Reworked the Barretenberg (async) and BarretenbergSync classes. They take options, one of which is an optional backend type. If not given will try and pick the best backend. Backends include:
  * Wasm
  * WasmWorker
  * NativeUnixDomainSocket
  * NativeSharedMemory
* There's a benchmark class in bb.js poseidon that measures cost for various backends. See results below. WASM is using the new msgpack api, Direct WASM is using the old cbind api (so you can see msgpack overhead).
* Only WASM and shared memory backends are supported by BarretenbergSync.
* All `foundation/crypto` primitives updated to use new api.
* **This PR does NOT enable native code.** To enable will require some ci efforts to copy `bb` and napi modules to `ts/build/<arch>-<os>` folder. Will be done in a follow up PR.

```
┌─ Size   2 field elements ──────────────────────────────────┐
│ WASM:                 491.61ms (  49.16µs/call) [baseline] │
│ Direct WASM:          438.22ms (  43.82µs/call) -  10.9%   │
│ Native Socket:        608.83ms (  60.88µs/call) +  23.8%   │
│ Native Shared:        268.16ms (  26.82µs/call) -  45.5%   │
│ Native Shared Sync:   238.86ms (  23.89µs/call) -  51.4%   │
└────────────────────────────────────────────────────────────┘
┌─ Size   4 field elements ──────────────────────────────────┐
│ WASM:                 738.71ms (  73.87µs/call) [baseline] │
│ Direct WASM:          667.25ms (  66.72µs/call) -   9.7%   │
│ Native Socket:        711.70ms (  71.17µs/call) -   3.7%   │
│ Native Shared:        418.90ms (  41.89µs/call) -  43.3%   │
│ Native Shared Sync:   383.39ms (  38.34µs/call) -  48.1%   │
└────────────────────────────────────────────────────────────┘
┌─ Size   8 field elements ──────────────────────────────────┐
│ WASM:                1066.59ms ( 106.66µs/call) [baseline] │
│ Direct WASM:          970.95ms (  97.09µs/call) -   9.0%   │
│ Native Socket:        870.64ms (  87.06µs/call) -  18.4%   │
│ Native Shared:        567.65ms (  56.77µs/call) -  46.8%   │
│ Native Shared Sync:   546.37ms (  54.64µs/call) -  48.8%   │
└────────────────────────────────────────────────────────────┘
```
